src/NUnitConsole/nunit-console/Utilities/TeamCityServiceMessagePublisher.cs

```diff
@@ -0,0 +1,389 @@
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.IO;
+using System.Xml;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NUnit.ConsoleRunner.Utilities
+{
+    using System;
+
+    internal class TeamCityServiceMessagePublisher
+    {
+        private readonly TextWriter _outWriter;
+        private readonly Dictionary<string, string> _refs = new Dictionary<string, string>();
+        private int _blockCounter;
+        private string _rootFlowId;
+
+        public TeamCityServiceMessagePublisher(TextWriter outWriter)
+        {
+            if (outWriter == null)
+            {
+                throw new ArgumentNullException("outWriter");
+            }
+
+            _outWriter = outWriter;
+        }
+
+        public void RegisterMessage(XmlNode message)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException("message");
+            }
+
+            var messageName = message.Name;
+            if (string.IsNullOrEmpty(messageName))
+            {
+                return;
+            }
+
+            messageName = messageName.ToLowerInvariant();
+            if (messageName == "start-run")
+            {
+                _refs.Clear();
+                return;
+            }           
+
+            var fullName = message.GetAttribute("fullname");
+            if (string.IsNullOrEmpty(fullName))
+            {
+                return;
+            }
+
+            var id = message.GetAttribute("id");
+            var parentId = message.GetAttribute("parentId");
+            string flowId;
+            if (parentId != null)
+            {
+                // NUnit 3 case
+                string rootId;
+                flowId = TryFindRootId(parentId, out rootId) ? rootId : id;
+            }
+            else
+            {
+                // NUnit 2 case
+                flowId = _rootFlowId;
+            }
+
+            string testFlowId;
+            if (id != flowId && parentId != null)
+            {
+                testFlowId = id;
+            }
+            else
+            {
+                testFlowId = flowId;
+                if (testFlowId == null)
+                {
+                    testFlowId = id;
+                }
+            }
```

Comment:
That's exactly right. I think we might need something like a worker id, which would replace the managed thread id. That's because we will dynamically create workers in the future and managed thread ids get reused at times. Also, it's possible that a worker thread is cancelled due to a problem and the worker is restarted on a new thread! Adding such an id would be easy enough.

There is definitely a problem with the hierarchical aspect. A SetUpFixture runs on some worker, then the fixtures run on various workers, finally the setUpFixture finishes on the original worker. However, SetUpFixtures don't contain tests, so maybe that's not a problem for you. OTOH, when we implement parallelism of cases inside a fixture, the problem returns big time.

If we provide parentid, does that actually eliminate the need for a separate flowid?